### PR TITLE
milestone integrations

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: consensus/algorithm.go [hunk 1]
```diff
@@ -1,44 +1,54 @@
...
+	//todo feed all layers from b's layer to tortoise
 	log.Info("received block with layer Id %v block id: %v ", b.Layer(), b.ID())
 }
 
-func (alg *Algorithm) HandleIncomingLayer(ll *mesh.Layer) {
+func (alg *Algorithm) HandleIncomingLayer(ll *mesh.Layer) (mesh.LayerID, mesh.LayerID) {
+	old := alg.latestComplete()
 	alg.Tortoise.handleIncomingLayer(ll)
-	alg.callback(ll.Index())
+	new := alg.latestComplete()
```
Comment (y0sher) [inline consensus/algorithm.go]: new is a lang keyword

File: consensus/ninja_tortoise.go [hunk 2]
```diff
@@ -116,7 +116,7 @@ func (ni *ninjaTortoise) processBlock(b *mesh.Block) {
 		ni.Debug("block votes %d", bid)
 		bl, found := ni.blocks[bid]
 		if !found {
-			panic("unknown block!, something went wrong ")
+			panic(fmt.Sprintf("error block not found ID %d", bid))
```
Comment (y0sher) [inline consensus/ninja_tortoise.go:119]: should we panic in real life ?

File: miner/block_builder.go [hunk 3]
```diff
@@ -123,15 +124,15 @@ func (t *BlockBuilder) AddTransaction(nonce uint64, origin, destination address.
 func (t *BlockBuilder) createBlock(id mesh.LayerID, txs []mesh.SerializableTransaction) mesh.Block {
 	var res []mesh.BlockID = nil
 	var err error
-	if id > 0 {
+	if id == 1 {
+		res = append(res, config.GenesisId)
+	} else {
 		res, err = t.hareResult.GetResult(id - 1)
```
Comment (antonlerner) [inline miner/block_builder.go:135]: get result will fail for id = 0

File: mesh/meshdb.go [hunk 4]
```diff
@@ -185,7 +185,7 @@ func (m *meshDB) handleLayerBlocks(ll *layerHandler) {
 
 			if err := m.blocks.Put(bl.ID().ToBytes(), bytes); err != nil {
 				log.Error("could not add bl to ", bl.ID(), " database ", err)
-				continue
+				return
```
Comment (antonlerner) [inline mesh/meshdb.go:188]: panic here, this should not happen, if we cant insert into DB we cant rely on our network image in meshdb anymore

File: mesh/mesh.go [hunk 5]
```diff
@@ -205,54 +190,54 @@ func (m *Mesh) handleOrphanBlocks(block *Block) {
...
+		panic(fmt.Sprint("failed getting latest layer  ", err))
+	}
+
+	//add last layer blocks
+	for _, b := range prevLayer.Blocks() {
+		ids[b.ID()] = struct{}{}
+	}
+
+	idArr := make([]BlockID, 0, len(ids))
+	for i := range ids {
```
Comment (antonlerner) [inline mesh/mesh.go:240]: we cant use a map to create this slice, the blocks must be in the same order in all nodes. iterating a map is random and therefore will not produce same results

File: mesh/mesh.go [hunk 6]
```diff
@@ -124,38 +109,37 @@ func (m *Mesh) AddLayer(layer *Layer) error {
...
-	sort.Slice(l.blocks, func(i, j int) bool {
-		return l.blocks[i].Id < l.blocks[j].Id
-	}) //todo: blocks are now sorted... what does it mean? does the ordering degrade security?
+		}
 
-	for _, b := range l.blocks {
-		for _, tx := range b.Txs {
-			//todo: think about these conversions.. are they needed?
-			txs = append(txs, SerializableTransaction2StateTransaction(&tx))
+		for _, b := range l.blocks {
```
Comment (antonlerner) [inline mesh/mesh.go:133]: please sort blocks, layer in db may not be in the same order between nodes

File: consensus/ninja_tortoise.go (hunk 2 above)
Comment (almogdepaz) [reply consensus/ninja_tortoise.go:119]: yup

**Review by antonlerner — APPROVED** [review]

**Pull request merged**